modules/standard/Set.chpl

```diff
@@ -143,59 +155,23 @@ module Set {
 
       :arg eltType: The type of the elements of this set.
       :arg parSafe: If `true`, this set will use parallel safe operations.
+      :arg resizeThreshold: Fractional value that specifies how full this map
+                            can be before requesting additional memory.
+      :arg initialCapacity: Integer value that specifies starting map size. The
+                            map can hold at least this many values before
+                            attempting to resize.
     */
-    proc init(type eltType, param parSafe=false) {
+    proc init(type eltType, param parSafe=false, resizeThreshold=0.5,
+              initialCapacity=32) {
       _checkElementType(eltType);
       this.eltType = eltType;
       this.parSafe = parSafe;
-      this._htb = new chpl__hashtable(eltType, nothing);
-    }
-
-    // Do things the slow/copy way if the element is serializable.
-    // See issue: #17477
-    pragma "no doc"
-    proc _addElem(in elem: eltType): bool
-    where _isSerializable(eltType) {
-        var result = true;
-
-        on this {
-          var (isFullSlot, idx) = _htb.findAvailableSlot(elem);
-
-          if !isFullSlot {
-            _htb.fillSlot(idx, elem, none);
-            result = true;
-          }
-        }
-
-      return result;
-    }
-
-    // For types that aren't serializable, avoid an extra copy by moving
-    // the value across locales.
-    pragma "no doc"
-    proc _addElem(pragma "no auto destroy" in elem: eltType): bool {
-      use Memory.Initialization;
-
-      var result = false;
-
-      on this {
-
-        // TODO: The following variation gets lifetime errors in
-        // '.../Set/types/testNilableTuple.chpl':
-        //
-        // var moved = moveToValue(elem);
-        // var (isFullSlot, idx) = _htb.findAvailableSlot(moved);
-        //
-        var (isFullSlot, idx) = _htb.findAvailableSlot(elem);
-
-        if !isFullSlot {
-          var moved = moveToValue(elem);
-          _htb.fillSlot(idx, moved, none);
-          result = true;
-        }
-      }
-
-      return result;
+      if boundsChecking then
+        if resizeThreshold <= 0 || resizeThreshold >= 1 then
+          boundsCheckHalt("'resizeThreshold' must be between 0 and 1");
```

Comment:
Hmm, that is a good point. I think part of the idea of bounds checking is you shouldn't get any halts when it is turned off ever, but I would need to check with Elliot about that.

Either way, the `map` module also does this boundsChecking check, so I might merge this one with it either way and then have a separate PR that removes that bounds check for both `set` and `map` if Elliot thinks they should be removed. Will consider this resolved unless you disagree with that. Thanks.

Reply:
I see, that makes sense.

I agree that consistency between the two is important.

Given that this is only in the init, I'd expect there to be low risk of programs not hitting the check in testing and hitting it with `--fast`, so that seems nice. I am a bit weary of not checking things, but I know halting from init is the only option right now I think.

Two ideas are
1) make `resizeThreshold` a param so you can do the bounds check at compile time always
2) without boundsChecking, clamp `resizeThreshold` to the range `[0, 1]`

I don't think either ideas are particularly great but just brainstorming.

So I think your plan is fine. It would be great to capture this discussion as it relates to the ongoing container parsafe questions (Though its not about parsafe, it feels related). In short: should containers also elide other bounds checks with `--fast`?

Reply:
Sounds good, I will log that in the parSafe main issue. I agree, would be good to capture this.

Reply:
Just spoke with Elliot about this, he said that we should switch to your clamping approach and then warn if the value gets clamped. 

Will implement that in a separate PR.